File: Example/macOS-Objective-C/ViewController.m
//
//  ViewController.m
//  macOS-Objective-C
//
//  Created by Felix Herrmann on 23.10.21.
//

#import "ViewController.h"
#import "PaintingItem.h"
#import "macOS_Objective_C-Swift.h"


@interface ViewController ()

@property NSScrollView* scrollView;
@property NSCollectionView* collectionView;

@property NSCollectionViewDiffableDataSource<NSString *,Painting *> *dataSource;

@end


@implementation ViewController


#pragma mark - ViewController

- (void)loadView {
    self.collectionView = [[NSCollectionView alloc] init];
    self.collectionView.collectionViewLayout = [self createLayout];
    [self.collectionView registerClass:PaintingItem.class forItemWithIdentifier:@"painting"];
    
    self.scrollView = [[NSScrollView alloc] initWithFrame:CGRectMake(0, 0, 800, 630)];
    self.scrollView.documentView = self.collectionView;
    self.view = self.scrollView;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    [self configureDataSource];
    [self applySnapshot];
}


#pragma mark - CollectionViewLayout

- (NSCollectionViewLayout *)createLayout {
    NSCollectionViewCompositionalLayout *layout = [[NSCollectionViewCompositionalLayout alloc] initWithSectionProvider:^NSCollectionLayoutSection * _Nullable(NSInteger _, id<NSCollectionLayoutEnvironment> _Nonnull environment) {
        CGFloat itemsPerWidth = (CGFloat)(int)(environment.container.effectiveContentSize.width / 300);
        CGFloat fractionalWidth = 1 / (itemsPerWidth < 1 ? 1 : itemsPerWidth);
        NSCollectionLayoutSize *itemSize = [NSCollectionLayoutSize sizeWithWidthDimension:[NSCollectionLayoutDimension fractionalWidthDimension:fractionalWidth] heightDimension:[NSCollectionLayoutDimension fractionalHeightDimension:1]];
        NSCollectionLayoutItem *item = [NSCollectionLayoutItem itemWithLayoutSize:itemSize];
        NSCollectionLayoutSize *groupSize = [NSCollectionLayoutSize sizeWithWidthDimension:[NSCollectionLayoutDimension fractionalWidthDimension:1]  heightDimension:[NSCollectionLayoutDimension absoluteDimension:100]];
        NSCollectionLayoutGroup *group = [NSCollectionLayoutGroup horizontalGroupWithLayoutSize:groupSize subitems:@[item]];
        NSCollectionLayoutSection *section = [NSCollectionLayoutSection sectionWithGroup:group];
        return section;
    }];
    return layout;
}


#pragma mark - DiffableDataSource

- (void)configureDataSource {
    self.dataSource = [[NSCollectionViewDiffableDataSource alloc] initWithCollectionView:self.collectionView itemProvider:^NSCollectionViewItem * _Nullable(NSCollectionView * _Nonnull collectionView, NSIndexPath * _Nonnull indexPath, Painting * _Nonnull painting) {
        PaintingItem *item = [collectionView makeItemWithIdentifier:@"painting" forIndexPath:indexPath];
        item.image = painting.image;
        item.mainLabel.text = painting.name;
        item.secondaryLabel.text = painting.artist;
        item.detailLabel.text = [NSString stringWithFormat:@"• %zd", painting.year];
        return item;
    }];
}

- (void)applySnapshot {
    NSDiffableDataSourceSnapshot<NSString *,Painting *> *snapshot = [[NSDiffableDataSourceSnapshot alloc] init];
    [snapshot appendSectionsWithIdentifiers:@[@"main"]];
    [snapshot appendItemsWithIdentifiers:Painting.examples intoSectionWithIdentifier:@"main"];
    [self.dataSource applySnapshot:snapshot animatingDifferences:NO];
}

@end

File: Example/macOS/ViewController.swift
//
//  ViewController.swift
//  macOS
//
//  Created by Felix Herrmann on 14.10.21.
//

import AppKit

final class ViewController: NSViewController {
    
    
    // MARK: - Properties
    
    private let scrollView = NSScrollView(frame: NSRect(x: 0, y: 0, width: 800, height: 630))
    private let collectionView = NSCollectionView()
    
    private lazy var dataSource: NSCollectionViewDiffableDataSource<String, Painting> = {
        return NSCollectionViewDiffableDataSource(collectionView: collectionView) { collectionView, indexPath, painting in
            let item = collectionView.makeItem(withIdentifier: .painting, for: indexPath) as? PaintingItem
            item?.image = painting.image
            item?.mainLabel.text = painting.name
            item?.secondaryLabel.text = painting.artist
            item?.detailLabel.text = "• \(painting.year)"
            return item
        }
    }()
    
    
    // MARK: - ViewController
    
    override func loadView() {
        collectionView.collectionViewLayout = createLayout()
        collectionView.register(PaintingItem.self, forItemWithIdentifier: .painting)
        
        scrollView.documentView = collectionView
        view = scrollView
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        applySnapshot()
    }
    
    
    // MARK: - CollectionViewLayout
    
    private func createLayout() -> NSCollectionViewLayout {
        let layout = NSCollectionViewCompositionalLayout { _, environment in
            let itemsPerWidth = CGFloat(Int(environment.container.effectiveContentSize.width / 300))
            let fractionalWidth = 1 / (itemsPerWidth < 1 ? 1 : itemsPerWidth)
            let itemSize = NSCollectionLayoutSize(widthDimension: .fractionalWidth(fractionalWidth), heightDimension: .fractionalHeight(1))
            let item = NSCollectionLayoutItem(layoutSize: itemSize)
            let groupSize = NSCollectionLayoutSize(widthDimension: .fractionalWidth(1), heightDimension: .absolute(100))
            let group = NSCollectionLayoutGroup.horizontal(layoutSize: groupSize, subitems: [item])
            let section = NSCollectionLayoutSection(group: group)
            return section
        }
        return layout
    }
    
    
    // MARK: - DiffableDataSource
    
    private func applySnapshot() {
        var snapshot = NSDiffableDataSourceSnapshot<String, Painting>()
        snapshot.appendSections(["main"])
        snapshot.appendItems(Painting.examples, toSection: "main")
        dataSource.apply(snapshot, animatingDifferences: false)
    }
}


// MARK: - NSUserInterfaceItemIdentifier + painting

extension NSUserInterfaceItemIdentifier {
    static let painting = NSUserInterfaceItemIdentifier("painting")
}
